cdk_eoapi/pgStacInfra.py

```diff
@@ -0,0 +1,167 @@
+import boto3
+
+
+class pgStacInfraStack(Stack):
+    def __init__(
+        self,
+        scope: Construct,
+        id: str,
+        vpc: aws_ec2.Vpc,
+        stage: str,
+        db_allocated_storage: int,
+        public_db_subnet: bool,
+        db_instance_type: str,
+        stac_api_lambda_name: str,
+        titiler_pgstac_api_lambda_name: str,
+        bastion_host_allow_ip_list: list,
+        bastion_host_create_elastic_ip: bool,
+        titiler_buckets: list,
+        data_access_role_arn: Optional[str],
+        auth_provider_jwks_url: Optional[str],
+        bastion_host_user_data: Union[str, aws_ec2.UserData],
+        **kwargs,
+    ) -> None:
+        super().__init__(scope, id, **kwargs)
+
+        pgstac_db = PgStacDatabase(
+            self,
+            "pgstac-db",
+            vpc=vpc,
+            engine=aws_rds.DatabaseInstanceEngine.postgres(
+                version=aws_rds.PostgresEngineVersion.VER_14
+            ),
+            vpc_subnets=aws_ec2.SubnetSelection(
+                subnet_type=aws_ec2.SubnetType.PUBLIC
+                if public_db_subnet
+                else aws_ec2.SubnetType.PRIVATE_ISOLATED
+            ),
+            allocated_storage=db_allocated_storage,
+            instance_type=aws_ec2.InstanceType(db_instance_type),
+        )
+
+        stac_api_lambda = PgStacApiLambda(
+            self,
+            "pgstac-api",
+            api_env=dict(NAME=stac_api_lambda_name, description=f"{stage} STAC API"),
+            vpc=vpc,
+            db=pgstac_db.db,
+            db_secret=pgstac_db.pgstac_secret,
+            subnet_selection=aws_ec2.SubnetSelection(
+                subnet_type=aws_ec2.SubnetType.PUBLIC
+                if public_db_subnet
+                else aws_ec2.SubnetType.PRIVATE_WITH_EGRESS
+            ),
+        )
+
+        titiler_pgstac_api_lambda = TitilerPgstacApiLambda(
+            self,
+            "titiler-pgstac-api",
+            api_env=dict(
+                NAME=titiler_pgstac_api_lambda_name,
+                description=f"{stage} titiler pgstac API",
+            ),
+            vpc=vpc,
+            db=pgstac_db.db,
+            db_secret=pgstac_db.pgstac_secret,
+            subnet_selection=aws_ec2.SubnetSelection(
+                subnet_type=aws_ec2.SubnetType.PUBLIC
+                if public_db_subnet
+                else aws_ec2.SubnetType.PRIVATE_WITH_EGRESS
+            ),
+            buckets=titiler_buckets,
+        )
+
+        bastion_host = BastionHost(
+            self,
+            "bastion-host",
+            vpc=vpc,
+            db=pgstac_db.db,
+            ipv4_allowlist=bastion_host_allow_ip_list,
+            user_data=aws_ec2.UserData.custom(bastion_host_user_data)
+            if bastion_host_user_data
+            else aws_ec2.UserData.for_linux(),
+            create_elastic_ip=bastion_host_create_elastic_ip,
+        )
+
+        if data_access_role_arn:
+            data_access_role = aws_iam.Role.from_role_arn(
```

Comment:
@emileten Don't we also need to apply `grant_assume_role` to a pre-existing `data_access_role` as well?  Did your initial tests include an injected `data _access_role`?

Reply:
@sharkinsspatial we've already talked about that, but 

If the user provides a data access role ARN, that role is going to be imported here, [and it's impossible to change the trust policy of an _imported_ role](https://github.com/aws/aws-cdk/discussions/23486#discussioncomment-4510475), so we must assume the role is already configured so that it can be assumed and used by the STAC ingestor. The only thing we can do is to warn the user about that. 

So, I thought we could leave things as is and write in the documentation of the `pydantic` `data_access_role_arn` field in `config.py` the following : 

> The STAC ingestor performs a set of validations of a STAC item being ingested. If the item has assets pointing to S3 objects, the STAC ingestor verifies these are accessible by _assuming_ the data access role provided and performing a HEAD object operation on the object. In order for this to work, two conditions relating to the data access role configuration must be verified : (1) the role has permissions to perform S3 READ operations and (2) the stac ingestor lambda role is authorized to assume the data access role. If the value of this parameter is `None`, a data access role is created as part of this deployment with a configuration verifying the above conditions (regarding (1), the role gets READ access on any S3 objects). However, if an ARN value is provided, the corresponding role is imported and assumed by the STAC ingestor and must thus have the right configuration before requests are sent to the ingestor. Below are examples of policies that would work for the default settings and can be manually added to the custom role on the AWS console  : 
> 
> 
> trust relationship policy 
> 
> 
> ```
> {
>     "Version": "2012-10-17",
>     "Statement": [
>         {
>             "Effect": "Allow",
>             "Principal": {
>                 "Service": "lambda.amazonaws.com"
>             },
>             "Action": "sts:AssumeRole"
>         },
>         {
>             "Effect": "Allow",
>             "Principal": {
>                 "AWS": "*"
>             },
>             "Action": "sts:AssumeRole",
>             "Condition": {
>                 "StringLike": {
>                     "aws:PrincipalArn": "arn:aws:iam::916098889494:role/*eoapi*stacingestor*"
>                 }
>             }
>         }
>     ]
> }
> ```
> 
> permission policy 
> 
> 
> ```
> {
>     "Version": "2012-10-17",
>     "Statement": [
>         {
>             "Effect": "Allow",
>             "Action": [
>                 "s3:Get*",
>                 "s3:List*",
>                 "s3-object-lambda:Get*",
>                 "s3-object-lambda:List*"
>             ],
>             "Resource": "*"
>         }
>     ]
> }
> ```
>